### ISSUE TITLE
False-positive ```missing-argument``` when instantiating DataTable2 in flet_datatable2

### ISSUE THREAD

**Issue by Dracovolodeus** [astral-sh/ty#3201]
### Summary

The ```ty``` analyzer incorrectly reports ```error[missing-argument]``` for the [DataTable2](https://github.com/flet-dev/flet/blob/main/sdk/python/packages/flet-datatable2/src/flet_datatable2/datatable2.py) dataclass. Specifically, it demands ```data_row_min_height``` and ```data_row_max_height```, despite both being explicitly defined with ```init=False``` via ```dataclasses.field```. While these fields exist in the parent class, they should be excluded from the constructor signature in the subclass.

#### Minimal reproduction:
```python
import flet as ft
import flet_datatable2 as fdt


def main(page: ft.Page):
    table = fdt.DataTable2(
        columns=[
            fdt.DataColumn2(label=ft.Text("First column")),
        ],
        rows=[
            fdt.DataRow2(
                cells=[
                    ft.DataCell(ft.Text("Data for first column"))
                ]
            )
        ],
    )
    page.add(table)


ft.app(main)
```
The code executes successfully at runtime.

#### Type checking this results in the following ```false-positive``` error:
```
> uv run ty --version
ty 0.0.27
> uv run ty check test.py
error[missing-argument]: No arguments provided for required parameters `data_row_min_height`, `data_row_max_height`
  --> test.py:6:13
   |
 5 |   def main(page: ft.Page):
 6 |       table = fdt.DataTable2(
   |  _____________^
 7 | |         columns=[
 8 | |             fdt.DataColumn2(label=ft.Text("First column")),
 9 | |         ],
10 | |         rows=[
11 | |             fdt.DataRow2(
12 | |                 cells=[
13 | |                     ft.DataCell(ft.Text("Data for first column"))
14 | |                 ]
15 | |             )
16 | |         ],
17 | |     )
   | |_____^
18 |       page.add(table)
   |
info: rule `missing-argument` is enabled by default

Found 1 diagnostic
```

#### Runtime inspection via ```inspect.signature``` further verifies that these parameters are not present in the class constructor.

<details>
<summary>Code</summary>

```python
import inspect
from pprint import pprint

from flet_datatable2 import DataTable2

pprint(list(inspect.signature(DataTable2.__init__).parameters.keys()))
```
</details>

<details>
<summary>Result</summary>

```
['self',
 'columns',
 'rows',
 'sort_ascending',
 'show_checkbox_column',
 'sort_column_index',
 'show_bottom_border',
 'border',
 'border_radius',
 'horizontal_lines',
 'vertical_lines',
 'checkbox_horizontal_margin',
 'column_spacing',
 'data_row_color',
 'data_text_style',
 'bgcolor',
 'gradient',
 'divider_thickness',
 'heading_row_color',
 'heading_row_height',
 'heading_text_style',
 'horizontal_margin',
 'clip_behavior',
 'on_select_all',
 'empty',
 'bottom_margin',
 'lm_ratio',
 'sm_ratio',
 'fixed_left_columns',
 'fixed_top_rows',
 'fixed_columns_color',
 'fixed_corner_color',
 'sort_arrow_icon_color',
 'min_width',
 'show_heading_checkbox',
 'heading_checkbox_theme',
 'data_row_checkbox_theme',
 'sort_arrow_icon',
 'sort_arrow_animation_duration',
 'visible_horizontal_scroll_bar',
 'visible_vertical_scroll_bar',
 'checkbox_alignment',
 'data_row_height',
 'data',
 'key',
 'ref',
 'expand',
 'expand_loose',
 'col',
 'opacity',
 'tooltip',
 'badge',
 'visible',
 'disabled',
 'rtl',
 'width',
 'height',
 'left',
 'top',
 'right',
 'bottom',
 'align',
 'margin',
 'rotate',
 'scale',
 'offset',
 'flip',
 'transform',
 'aspect_ratio',
 'animate_opacity',
 'animate_size',
 'animate_position',
 'animate_align',
 'animate_margin',
 'animate_rotation',
 'animate_scale',
 'animate_offset',
 'size_change_interval',
 'on_size_change',
 'on_animation_end']
```
</details>

#### My full pyproject.toml uv venv looks like this:

<details>
<summary>pyproject.toml</summary>

```toml
[project]
name = "ty-test"
version = "0.1.0"
description = "ty false-positive error demo"
readme = "README.md"
requires-python = ">=3.14"
dependencies = [
    "ty>=0.0.27",
    "flet-datatable2>=0.83.1",
    "flet[all]>=0.83.1",
]
```
</details>

### Version

ty 0.0.27

**Comment by sharkdp** [comment]
Interesting, thank you for reporting this.

`DataTable2` is decorated with a dataclass-transformer called `ft.control` which is defined here:

https://github.com/flet-dev/flet/blob/189472f0c825f01559e2e96cd3b93e33ae85faba/sdk/python/packages/flet/src/flet/controls/base_control.py#L89-L96

`ft.control` does not declare any `field_specifiers` in the `dataclass_transform` call. If I am reading the typing spec correctly, that means that *no field specifier* is supported:

> `field_specifiers` specifies a static list of supported classes that describe fields. Some libraries also supply functions to allocate instances of field specifiers, and those functions may also be specified in this tuple. **If not specified, `field_specifiers` will default to an empty tuple (no field specifiers supported)**. The standard dataclass behavior supports only one type of field specifier called `Field` plus a helper function (`field`) that instantiates this class, so if we were describing the stdlib dataclass behavior, we would provide the tuple argument `(dataclasses.Field, dataclasses.field)`.

So this looks to me like a bug in `flet`. If they want `dataclasses.field` calls to be respected, they should change [this line](https://github.com/flet-dev/flet/blob/189472f0c825f01559e2e96cd3b93e33ae85faba/sdk/python/packages/flet/src/flet/controls/base_control.py#L89) to
```py
@dataclass_transform(field_specifiers=(Field, field))
```
(and import `Field` from `dataclasses` as well)

If I make that change locally, ty runs fine on your example.

**Comment by carljm** [comment]
I explored this a bit, thinking that other type checkers were wrongly treating `dataclasses.field` as a field specifier, even when not listed explicitly, and that I'd want to raise this on discuss.python.org. But this does not seem to be the case -- only zuban seems to actually make this mistake. What is happening for other type checkers is more subtle.

Consider this simplified example:

```py
from dataclasses import field
from typing import dataclass_transform

@dataclass_transform()
def mydc[T](cls: type[T]) -> type[T]:
    raise NotImplementedError()

@mydc
class Base:
    hidden: None = field(init=False)

Base()
```

Mypy/pyright/pyrefly do not wrongly treat `field(...)` as a field specifier here; they simply treat it as an unknown default value for the field. Since `field` is annotated in typeshed as returning `T` (where `T` is the type of `default`, or the type returned from `default_factory`, and thus resolves to `Unknown/Any` if the call to `field` doesn't give a `default` or `default_factory`), the "wrong default value" (e.g. `x: int = field(...)` when `field` is not a field specifier) passes silently, even when `field` is not listed in `dataclass_specifiers` (this part is unfortunate and I would like to start a discussion about changing these typeshed annotations). The field is then included in the generated `__init__`, but with a default, so the call `Base()` still succeeds without providing that argument.

ty has a bug here. We currently partially special-case `dataclasses.field`  (without properly respecting `field_specifiers`) in a way which causes us to treat `hidden` as not having a default value, and thus we wrongly error on the `Base()` call above.

**Comment by carljm** [comment]
To be clear, this is still a bug in flet, and they should still make the change @sharkdp recommends above. It's just that, given the current (IMO wrong) annotations for `dataclasses.field` in typeshed, with a correctly-behaving type checker this bug will pass silently, causing the field to be treated as having a default, rather than as being `init=False`. ty highlights the bug in flet because of our own bug :)

**Comment by carljm** [comment]
https://discuss.python.org/t/can-we-have-a-more-truthful-typeshed-annotation-for-dataclasses-field/106832

**Comment by carljm** [comment]
https://github.com/flet-dev/flet/issues/6377
